src/common/mempool.cc

```diff
@@ -95,9 +97,11 @@ size_t mempool::pool_t::allocated_items() const
 
 void mempool::pool_t::adjust_count(ssize_t items, ssize_t bytes)
 {
+#ifndef WITH_SEASTAR
   thread_shard_index = (thread_shard_index == num_shards) ? pick_a_shard_int() : thread_shard_index;
```

Comment:
This might be the single place that breaks the principle of shard-not-folloing-the-core change.

Reply:
Huh, we've gotten this quite recently, in 2022.

```
commit ddb493f63ec2f2521ba3f511ba894fcc55190cd7
Author: liqiang <liqiang64@huawei.com>
Date:   Thu Jan 6 16:09:20 2022 +0800

    Use thread-local pointer variables to save the shard location required by this thread.
    
    Signed-off-by: liqiang <liqiang64@huawei.com>

diff --git a/src/common/mempool.cc b/src/common/mempool.cc
index e4ab5d4a10a..669ea192167 100644
--- a/src/common/mempool.cc
+++ b/src/common/mempool.cc
@@ -15,7 +15,7 @@
 #include "include/mempool.h"
 #include "include/demangle.h"
 
-
+static thread_local mempool::shard_t *thread_shard = NULL;
 // default to debug_mode off
 bool mempool::debug_mode = false;
 
@@ -92,9 +92,9 @@ size_t mempool::pool_t::allocated_items() const
 
 void mempool::pool_t::adjust_count(ssize_t items, ssize_t bytes)
 {
-  shard_t *shard = pick_a_shard();
-  shard->items += items;
-  shard->bytes += bytes;
+  thread_shard = (thread_shard == NULL) ? pick_a_shard() : thread_shard;
+  thread_shard->items += items;
+  thread_shard->bytes += bytes;
 }
 
 void mempool::pool_t::get_stats(
```

Reply:
In non-jumping-across-cores workloads this optimization saves a look-up BUT **the benefit has been given up later**:

```
commit 6554fb615ad3f2c7b26a2e5e10dc12401be5836e
Author: liqiang <liqiang64@huawei.com>
Date:   Fri Jan 7 11:03:48 2022 +0800

    Thread local variables should hold the index of the thread, not the shard[] pointer of the class
    
    Signed-off-by: liqiang <liqiang64@huawei.com>

diff --git a/src/common/mempool.cc b/src/common/mempool.cc
index 70b58635067..79354f70821 100644
--- a/src/common/mempool.cc
+++ b/src/common/mempool.cc
@@ -15,7 +15,10 @@
 #include "include/mempool.h"
 #include "include/demangle.h"
 
-static thread_local mempool::shard_t *thread_shard = nullptr;
+// Thread local variables should save index, not &shard[index],
+// because shard[] is defined in the class
+static thread_local size_t thread_shard_index = mempool::num_shards;
+
 // default to debug_mode off
 bool mempool::debug_mode = false;
 
@@ -92,9 +95,9 @@ size_t mempool::pool_t::allocated_items() const
 
 void mempool::pool_t::adjust_count(ssize_t items, ssize_t bytes)
 {
-  thread_shard = (thread_shard == nullptr) ? pick_a_shard() : thread_shard;
-  thread_shard->items += items;
-  thread_shard->bytes += bytes;
+  thread_shard_index = (thread_shard_index == num_shards) ? pick_a_shard_int() : thread_shard_index;
+  shard[thread_shard_index].items += items;
+  shard[thread_shard_index].bytes += bytes;
 }
 
 void mempool::pool_t::get_stats(
```

I'm going to revert as this doesn't make sense anymore while affecting the classical world as well.

Reply:
> This might be the single place that breaks the shard-change-following-the-core-change principle.

I've made a thinko. This caching is based on a `thread_local`, so it actually honors the shard-change-follows-the-core-change rule (assuming threads are "glued" to cores; ideally it would per-cpu sharding); please ignore the comments above.

Hmm, just collisions on the shard selection then?